File: project/src/main/java/com/project/project/businesslogic/service/BookingService.java
package com.project.project.businesslogic.service;

import com.project.project.dataaccess.dao.BookingDao;
import com.project.project.entity.Booking;
import com.project.project.entity.Movie;
import com.project.project.entity.Show;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class BookingService {
    @Autowired
    private BookingDao bookingDao;

    public Booking addBooking(Booking booking) {
        return bookingDao.save(booking);
    }

    public List<Booking> getBookings() {
        return (List<Booking>) bookingDao.findAll();
    }

    public Booking getById(Long id) {
        return bookingDao.findById(id).get();
    }

    public void deleteById(Long id) {
        bookingDao.deleteById(id);
    }

    public List<Booking> getBookingsByUsername(String username) {
        return bookingDao.findByUsername(username);
    }

    public List<Movie> getMoviesByBookingId(Long id) {
        return bookingDao.findMoviesByBookingId(id);
    }

    public List<Show> getShowsByBookingId(Long id) {
        return bookingDao.findShowsByBookingId(id);
    }
}

File: project-frontend/src/app/service/booking.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Booking } from '../_model/booking';
import { Movie } from '../_model/movie';
import { Show } from '../_model/show';

@Injectable({
  providedIn: 'root'
})
export class BookingService {

  constructor(private httpClient: HttpClient) { }

  addBooking(booking: Booking, username: string, showId: number, movieId: number) {
    return this.httpClient.post<Booking>("http://localhost:8080/addBooking/" + username + '/' + showId + '/' + movieId + '/save', booking);
  }

  getBookings(username: string) {
    return this.httpClient.get<Booking[]>("http://localhost:8080/getBookingsByUsername/" + username);
  }

  getMoviesByBookingId(id: number) {
    return this.httpClient.get<Movie[]>("http://localhost:8080/getMoviesByBookingId/" + id);
  }

  getShowsByBookingId(id: number) {
    return this.httpClient.get<Show[]>("http://localhost:8080/getShowsByBookingId/" + id);
  }
}
